### PR TITLE
[#3471] Add Pub/Sub based Command & Control API

### PR DESCRIPTION
**Pull request by michelleFranke**
This is another PR to use Pub/Sub as only messaging infrastructure. As part of this architecture we updated the documentation for a Pub/Sub based Command & Control API. In a second commit we will provide the new module “command-pubsub” which includes a Pub/Sub based implementation of the client for Hono’s Command & Control API.

- Updated documentation.

### CODE REVIEW

File: site/documentation/content/api/command-and-control-pubsub/index.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+   from *Pub/Sub* and forwards it to the device, provided that the target device is connected and is accepting commands.
+
+**Message Format**
+
+Metadata MUST be set as Pub/Sub attributes on a message. The following table provides an overview of the attributes the
+*Business Application* needs to set on a one-way command message.
+
+| Name       | Mandatory | Type     | Description                                                   |
+|:-----------|:---------:|:---------|:--------------------------------------------------------------|
+| *deviceId* |    yes    | *string* | The identifier of the device that the command is targeted at. |
```
Comment (sophokles73) [inline site/documentation/content/api/command-and-control-pubsub/index.md:63]: we do not use camel case notation for any of the headers/attributes. Can you please change this to either *device-id* or (in analogy to the Kafka based API) *device_id*?

File: site/documentation/content/api/command-and-control-pubsub/index.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+| *tenantId* |    yes    | *string* | The identifier of the tenant that the command is targeted at. |
```
Comment (sophokles73) [inline site/documentation/content/api/command-and-control-pubsub/index.md:64]: is there a particular reason why this is needed? The tenant should be easily determined from the topic name, shouldn't it?

File: site/documentation/content/api/command-and-control-pubsub/index.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+| *subject*  |    yes    | *string* | The name of the command to be executed by the device.         |
```
Comment (sophokles73) [inline site/documentation/content/api/command-and-control-pubsub/index.md:65]: can we also please add an optional *content-type* attribute?

File: site/documentation/content/api/command-and-control-pubsub/index.md [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+**Command Message Format**
+
+Metadata MUST be set as Pub/Sub attributes on a message.
+The following table provides an overview of the attributes the *Business Application* needs to set on a request/response
+command message.
+
+| Name                                         | Mandatory | Type      | Description                                                                                                                                |
+|:---------------------------------------------|:---------:|:----------|:-------------------------------------------------------------------------------------------------------------------------------------------|
+| *correlation-id*                             |    yes    | *string*  | The identifier used to correlate a response message to the original request. It is used as the *correlation-id* attribute in the response. |
+| *deviceId*                                   |    yes    | *string*  | The identifier of the device that the command is targeted at.                                                                              |
```
Comment (sophokles73) [inline site/documentation/content/api/command-and-control-pubsub/index.md:118]: *device-id* or *device_id*

File: site/documentation/content/api/command-and-control-pubsub/index.md [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+| *response-required*                          |    yes    | *boolean* | MUST be set with a value of `true`, meaning that the device is required to send a response for the command.                                |
+| *subject*                                    |    yes    | *string*  | The name of the command to be executed by the device.                                                                                      |
+| *tenantId*                                   |    yes    | *string*  | The identifier of the tenant that the command is targeted at.                                                                              |
```
Comment (sophokles73) [inline site/documentation/content/api/command-and-control-pubsub/index.md:121]: is this required?

File: site/documentation/content/api/command-and-control-pubsub/index.md [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+|:---------------------------------------------|:---------:|:----------|:-------------------------------------------------------------------------------------------------------------------------------------------|
+| *correlation-id*                             |    yes    | *string*  | The identifier used to correlate a response message to the original request. It is used as the *correlation-id* attribute in the response. |
+| *deviceId*                                   |    yes    | *string*  | The identifier of the device that the command is targeted at.                                                                              |
+| *response-required*                          |    yes    | *boolean* | MUST be set with a value of `true`, meaning that the device is required to send a response for the command.                                |
+| *subject*                                    |    yes    | *string*  | The name of the command to be executed by the device.                                                                                      |
```
Comment (sophokles73) [inline site/documentation/content/api/command-and-control-pubsub/index.md:120]: can we also add an optional *content-type* attribute and *delivery-failure-notification-metadata* as defined by the Kafka based API?

File: site/documentation/content/api/command-and-control-pubsub/index.md [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+| *deviceId*       |    yes    | *string*  | The identifier of the device that sent the response.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
```
Comment (sophokles73) [inline site/documentation/content/api/command-and-control-pubsub/index.md:140]: *device-id* or *device_id*

File: site/documentation/content/api/command-and-control-pubsub/index.md [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+| *tenantId*       |    yes    | *string*  | The identifier of the tenant that the device belongs to.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
```
Comment (sophokles73) [inline site/documentation/content/api/command-and-control-pubsub/index.md:142]: *tenant-id* or *tenant_id*

File: site/documentation/content/api/command-and-control-pubsub/index.md [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+| *content-type*   |    no     | *string*  | If present, MUST contain a *Media Type* as defined by [RFC 2046](https://tools.ietf.org/html/rfc2046) which describes the semantics and format of the command's input data contained in the message payload. However, not all protocol adapters will support this property as not all transport protocols provide means to convey this information, e.g. MQTT 3.1.1 has no notion of message headers.<br>If the response is an error message sent by the Hono protocol adapter or Command Router component, the content type MUST be *application/vnd.eclipse-hono-delivery-failure-notification+json*.   |
```
Comment (sophokles73) [inline site/documentation/content/api/command-and-control-pubsub/index.md:143]: can we also please add *delivery-failure-notification-metadata* as defined by the Kafka based API?

**Review by sophokles73 — CHANGES_REQUESTED** [review]

File: site/documentation/content/api/command-and-control-pubsub/index.md (hunk 1 above)
Comment (michelleFranke) [reply site/documentation/content/api/command-and-control-pubsub/index.md]: sure, updated to device_id

File: site/documentation/content/api/command-and-control-pubsub/index.md (hunk 2 above)
Comment (michelleFranke) [reply site/documentation/content/api/command-and-control-pubsub/index.md]: no you're right, it can be determined from the topic name. Removed it

**Review by sophokles73 — APPROVED** [review]
LGTM

**Pull request merged**